json object access by index not parsing

Thanks for fixing the last property accessor bug. Here's another one. :)

Given:
```
$ cat test.bicep 
var env = 'dev'

var id = json(loadTextContent('./vars.json')).subscriptions[env].id

output testId string = id

$ cat vars.json 
{
  "subscriptions": {
    "prod": {
      "id": "an id"
    },
    "dev": {
      "id": "another id"
    }
  }
}
```
We get unexpected left side square bracket:
```
$ checkov -f test.bicep  --framework bicep                                                                                                                                                                     
2022-05-30 08:44:14,600 [MainThread  ] [ERROR]  [bicep] Couldn't parse test.bicep                                                                                                                                                           
Traceback (most recent call last):                                                                                                                                                                                                          
  File "/home/user/.local/lib/python3.10/site-packages/lark/parsers/lalr_parser.py", line 126, in feed_token                                                                                                                            
    action, arg = states[state][token.type]                                                                                                                                                                                                 
KeyError: 'LSQB'                                                                                                                                                                                                                            
                                                                                                                                                                                                                                            
During handling of the above exception, another exception occurred:                                                                                                                                                                         
                                                                                                                                                                                                                                            
Traceback (most recent call last):                                                                                                                                                                                                          
  File "/home/user/.local/lib/python3.10/site-packages/checkov/bicep/parser.py", line 20, in parse                                                                                                                                      
    template = self.bicep_parser.parse(text=content)                                                                                                                                                                                        
  File "/home/user/.local/lib/python3.10/site-packages/pycep/main.py", line 28, in parse                                                                                                                                                
    tree = self._create_tree(text=text, file_path=file_path)                                                                                                                                                                                
  File "/home/user/.local/lib/python3.10/site-packages/pycep/main.py", line 46, in _create_tree                                                                                                                                         
    return self.lark_parser.parse(bicep_text)                                                                                                                                                                                               
  File "/home/user/.local/lib/python3.10/site-packages/lark/lark.py", line 625, in parse                                                                                                                                                
    return self.parser.parse(text, start=start, on_error=on_error)                                                                                                                                                                          
  File "/home/user/.local/lib/python3.10/site-packages/lark/parser_frontends.py", line 96, in parse                                                                                                                                     
    return self.parser.parse(stream, chosen_start, **kw)                                                                                                                                                                                    
  File "/home/user/.local/lib/python3.10/site-packages/lark/parsers/lalr_parser.py", line 41, in parse
    return self.parser.parse(lexer, start)
  File "/home/user/.local/lib/python3.10/site-packages/lark/parsers/lalr_parser.py", line 171, in parse
    return self.parse_from_state(parser_state)
  File "/home/user/.local/lib/python3.10/site-packages/lark/parsers/lalr_parser.py", line 188, in parse_from_state 
    raise e
  File "/home/user/.local/lib/python3.10/site-packages/lark/parsers/lalr_parser.py", line 179, in parse_from_state 
    state.feed_token(token)
  File "/home/user/.local/lib/python3.10/site-packages/lark/parsers/lalr_parser.py", line 129, in feed_token
    raise UnexpectedToken(token, expected, state=self, interactive_parser=None)
lark.exceptions.UnexpectedToken: Unexpected token Token('LSQB', '[') at line 3, column 60.
```
These may come in any sane or insane order and still be valid, i.e:
```
var sub = 'subscriptions'
var var_id = 'id'

var id = json(loadTextContent('./vars.json'))[sub].dev[var_id]
```

--- a/.github/workflows/codeql-analysis.yml
+++ b/.github/workflows/codeql-analysis.yml
@@ -33,8 +33,8 @@ jobs:
       - name: Checkout repository
         uses: actions/checkout@2541b1294d2704b0964813337f33b291d3f8596b  # v3.0.2
       - name: Initialize CodeQL
-        uses: github/codeql-action/init@a3a6c128d771b6b9bdebb1c9d0583ebd2728a108  # v2.1.11
+        uses: github/codeql-action/init@27ea8f8fe5977c00f5b37e076ab846c5bd783b96  # v2.1.12
         with:
           languages: 'python'
       - name: Perform CodeQL Analysis
-        uses: github/codeql-action/analyze@a3a6c128d771b6b9bdebb1c9d0583ebd2728a108  # v2.1.11
+        uses: github/codeql-action/analyze@27ea8f8fe5977c00f5b37e076ab846c5bd783b96  # v2.1.12

--- a/.github/workflows/scorecards-analysis.yml
+++ b/.github/workflows/scorecards-analysis.yml
@@ -22,7 +22,7 @@ jobs:
         with:
           persist-credentials: false
       - name: "Run analysis"
-        uses: ossf/scorecard-action@5c8bc69dc88b65c66584e07611df79d3579b0377  # v1.1.0
+        uses: ossf/scorecard-action@3e15ea8318eee9b333819ec77a36aca8d39df13e  # v1.1.1
         with:
           results_file: results.sarif
           results_format: sarif
@@ -35,6 +35,6 @@ jobs:
           path: results.sarif
           retention-days: 7
       - name: "Upload to code-scanning"
-        uses: github/codeql-action/upload-sarif@a3a6c128d771b6b9bdebb1c9d0583ebd2728a108  # v2.1.11
+        uses: github/codeql-action/upload-sarif@27ea8f8fe5977c00f5b37e076ab846c5bd783b96  # v2.1.12
         with:
           sarif_file: results.sarif

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -9,6 +9,8 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 ### Added
 
+- Support index accessor for `json` function
+
 ## [0.3.6] - 2022-05-21
 
 ### Added
@@ -108,7 +110,7 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 - First release to be able to parse all the official examples of Bicep [101](https://github.com/Azure/bicep/tree/main/docs/examples/101).
 
 [Unreleased]: https://github.com/gruebel/pycep/compare/0.3.6...HEAD
-[0.3.5]: https://github.com/gruebel/pycep/compare/0.3.5...0.3.6
+[0.3.6]: https://github.com/gruebel/pycep/compare/0.3.5...0.3.6
 [0.3.5]: https://github.com/gruebel/pycep/compare/0.3.4...0.3.5
 [0.3.4]: https://github.com/gruebel/pycep/compare/0.3.3...0.3.4
 [0.3.3]: https://github.com/gruebel/pycep/compare/0.3.2...0.3.3

--- a/pycep/bicep.lark
+++ b/pycep/bicep.lark
@@ -175,7 +175,7 @@ int_func: ("int" | "sys.int") "(" value_brackets ")"
 
 ?function_object: json_func
 
-json_func: ("json" | "sys.json") "(" value_brackets ")" ["." STRING]
+json_func: ("json" | "sys.json") "(" value_brackets ")" [["[" STRING "]"] "." (STRING | INDEXED)]
 
 // functions - resource
 
@@ -183,7 +183,7 @@ json_func: ("json" | "sys.json") "(" value_brackets ")" ["." STRING]
 
 extension_resource_id: ("extensionResourceId" | "az.extensionResourceId") "(" value_brackets ("," value_brackets)~2..3 ")"
 
-list_keys: ("listKeys" | "az.listKeys") "(" value_brackets "," value_brackets ")" ["." (STRING | INDEXED)]
+list_keys: ("listKeys" | "az.listKeys") "(" value_brackets "," value_brackets ")" [["[" STRING "]"] "." (STRING | INDEXED)]
 
 pick_zones: ("pickZones" | "az.pickZones") "(" value_brackets ("," value_brackets)~2..4 ")"
 

--- a/pycep/transformer.py
+++ b/pycep/transformer.py
@@ -780,8 +780,10 @@ class BicepToJson(Transformer[Token, pycep_typing.BicepJson]):
     #
     ####################
 
-    def json_func(self, args: tuple[pycep_typing.PossibleValue, pycep_typing.PossibleNoneValue]) -> pycep_typing.Json:
-        arg_1, property_name = args
+    def json_func(
+        self, args: tuple[pycep_typing.PossibleValue, pycep_typing.PossibleNoneValue, pycep_typing.PossibleNoneValue]
+    ) -> pycep_typing.Json:
+        arg_1, property_name_1, property_name_2 = args
 
         result: pycep_typing.Json = {
             "type": "json",
@@ -790,8 +792,14 @@ class BicepToJson(Transformer[Token, pycep_typing.BicepJson]):
             },
         }
 
-        if property_name:
-            result["property_name"] = str(property_name)
+        if property_name_2:
+            # workaround for direct index accessors
+            if property_name_1:
+                property_name = f"[{property_name_1}]{property_name_2}"
+            else:
+                property_name = str(property_name_2)
+
+            result["property_name"] = property_name
 
         return result
 
@@ -825,9 +833,15 @@ class BicepToJson(Transformer[Token, pycep_typing.BicepJson]):
         }
 
     def list_keys(
-        self, args: tuple[pycep_typing.PossibleValue, pycep_typing.PossibleValue, pycep_typing.PossibleNoneValue]
+        self,
+        args: tuple[
+            pycep_typing.PossibleValue,
+            pycep_typing.PossibleValue,
+            pycep_typing.PossibleNoneValue,
+            pycep_typing.PossibleNoneValue,
+        ],
     ) -> pycep_typing.ListKeys:
-        resource_identifier, api_version, property_name = args
+        resource_identifier, api_version, property_name_1, property_name_2 = args
 
         result: pycep_typing.ListKeys = {
             "type": "list_keys",
@@ -837,7 +851,13 @@ class BicepToJson(Transformer[Token, pycep_typing.BicepJson]):
             },
         }
 
-        if property_name:
+        if property_name_2:
+            # workaround for direct index accessors
+            if property_name_1:
+                property_name = f"[{property_name_1}]{property_name_2}"
+            else:
+                property_name = str(property_name_2)
+
             result["property_name"] = str(property_name)
 
         return result

--- a/tests/test_function/examples/object/json/main.bicep
+++ b/tests/test_function/examples/object/json/main.bicep
@@ -21,5 +21,9 @@ resource identityProvider 'Microsoft.ApiManagement/service/identityProviders@202
 // output
 output ftpUser string = json(site.properties).id
 
+output lateIndexed string = json(loadTextContent('./vars.json')).subscriptions[env].id
+
+output directIndexed string = json(loadTextContent('./vars.json'))[sub].dev[var_id]
+
 // namespace sys
 output namespace string = sys.json(site)

--- a/tests/test_function/examples/object/json/result.json
+++ b/tests/test_function/examples/object/json/result.json
@@ -70,6 +70,48 @@
         }
       }
     },
+    "lateIndexed": {
+      "decorators": [],
+      "type": "string",
+      "value": {
+        "function": {
+          "type": "json",
+          "parameters": {
+            "arg_1": {
+              "function": {
+                "type": "load_text_content",
+                "parameters": {
+                  "file_path": "./vars.json",
+                  "encoding": null
+                }
+              }
+            }
+          },
+          "property_name": "subscriptions[env].id"
+        }
+      }
+    },
+    "directIndexed": {
+      "decorators": [],
+      "type": "string",
+      "value": {
+        "function": {
+          "type": "json",
+          "parameters": {
+            "arg_1": {
+              "function": {
+                "type": "load_text_content",
+                "parameters": {
+                  "file_path": "./vars.json",
+                  "encoding": null
+                }
+              }
+            }
+          },
+          "property_name": "[sub]dev[var_id]"
+        }
+      }
+    },
     "namespace": {
       "decorators": [],
       "type": "string",

